File: VulkanContext/Swapchain.cpp
#include "Swapchain.h"
#include "VkCheck.h"
#include <limits>
#include <algorithm>

namespace {
	VkSurfaceFormatKHR chooseSwapSurfaceFormat(const std::vector<VkSurfaceFormatKHR>& availableFormats)
	{
		for (const auto& availableFormat : availableFormats)
		{
			if (availableFormat.format == VK_FORMAT_B8G8R8A8_SRGB && availableFormat.colorSpace == VK_COLOR_SPACE_SRGB_NONLINEAR_KHR)
			{
				return availableFormat;
			}
		}
		return availableFormats[0];
	}

	VkPresentModeKHR chooseSwapPresentMode(const std::vector<VkPresentModeKHR>& availablePresentModes)
	{
		for (const auto& availablePresentMode : availablePresentModes)
		{
			if (availablePresentMode == VK_PRESENT_MODE_MAILBOX_KHR)
			{
				return availablePresentMode;
			}
		}
		return VK_PRESENT_MODE_FIFO_KHR;
	}

	VkExtent2D chooseSwapExtent(const VkSurfaceCapabilitiesKHR& capabilities, SDL_Window* window)
	{
		if (capabilities.currentExtent.width != std::numeric_limits<uint32_t>::max())
		{
			return capabilities.currentExtent;
		}

		int width, height;
		SDL_GetWindowSizeInPixels(window, &width, &height);

		VkExtent2D actualExtent = { static_cast<uint32_t>(width), static_cast<uint32_t>(height) };
		actualExtent.width = std::clamp(actualExtent.width, capabilities.minImageExtent.width, capabilities.maxImageExtent.width);
		actualExtent.height = std::clamp(actualExtent.height, capabilities.minImageExtent.height, capabilities.maxImageExtent.height);
		return actualExtent;
	}
}

Swapchain::Swapchain(Device& device, VkSurfaceKHR surface, SDL_Window* window)
	: device(device), surface(surface), window(window)
{
	createSwapchain();
	createImageViews();
}

Swapchain::~Swapchain()
{
	cleanup();
}

void Swapchain::recreate()
{
	int width = 0, height = 0;
	SDL_GetWindowSizeInPixels(window, &width, &height);
	while (width == 0 || height == 0)
	{
		SDL_GetWindowSizeInPixels(window, &width, &height);
		SDL_WaitEvent(nullptr);
	}

	vkDeviceWaitIdle(device.get());
	cleanup();
	createSwapchain();
	createImageViews();
}

void Swapchain::cleanup()
{
	for (auto imageView : imageViews_)
	{
		vkDestroyImageView(device.get(), imageView, nullptr);
	}
	vkDestroySwapchainKHR(device.get(), swapchain, nullptr);
}

void Swapchain::createSwapchain()
{
	SwapChainSupportDetails support = device.querySwapchainSupport(surface);
	VkSurfaceFormatKHR surfaceFormat = chooseSwapSurfaceFormat(support.formats);
	VkPresentModeKHR presentMode = chooseSwapPresentMode(support.presentModes);
	VkExtent2D extent = chooseSwapExtent(support.capabilities, window);

	uint32_t imageCount = support.capabilities.minImageCount + 1;
	if (support.capabilities.maxImageCount > 0 && imageCount > support.capabilities.maxImageCount)
	{
		imageCount = support.capabilities.maxImageCount;
	}

	VkSwapchainCreateInfoKHR createInfo{};
	createInfo.sType = VK_STRUCTURE_TYPE_SWAPCHAIN_CREATE_INFO_KHR;
	createInfo.surface = surface;
	createInfo.minImageCount = imageCount;
	createInfo.imageFormat = surfaceFormat.format;
	createInfo.imageColorSpace = surfaceFormat.colorSpace;
	createInfo.imageExtent = extent;
	createInfo.imageArrayLayers = 1;
	createInfo.imageUsage = VK_IMAGE_USAGE_COLOR_ATTACHMENT_BIT;

	uint32_t queueFamilyIndices[] = { device.graphicsFamily(), device.presentFamily() };
	if (device.graphicsFamily() != device.presentFamily())
	{
		createInfo.imageSharingMode = VK_SHARING_MODE_CONCURRENT;
		createInfo.queueFamilyIndexCount = 2;
		createInfo.pQueueFamilyIndices = queueFamilyIndices;
	}
	else
	{
		createInfo.imageSharingMode = VK_SHARING_MODE_EXCLUSIVE;
	}

	createInfo.preTransform = support.capabilities.currentTransform;
	createInfo.compositeAlpha = VK_COMPOSITE_ALPHA_OPAQUE_BIT_KHR;
	createInfo.presentMode = presentMode;
	createInfo.clipped = VK_TRUE;
	createInfo.oldSwapchain = VK_NULL_HANDLE;

	VK_CHECK(vkCreateSwapchainKHR(device.get(), &createInfo, nullptr, &swapchain));
	vkGetSwapchainImagesKHR(device.get(), swapchain, &imageCount, nullptr);
	images_.resize(imageCount);
	vkGetSwapchainImagesKHR(device.get(), swapchain, &imageCount, images_.data());

	imageFormat_ = surfaceFormat.format;
	extent_ = extent;
}

void Swapchain::createImageViews()
{
	imageViews_.resize(images_.size());

	for (size_t i = 0; i < images_.size(); i++)
	{
		VkImageViewCreateInfo createInfo{};
		createInfo.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
		createInfo.image = images_[i];
		createInfo.viewType = VK_IMAGE_VIEW_TYPE_2D;
		createInfo.format = imageFormat_;
		createInfo.components.r = VK_COMPONENT_SWIZZLE_IDENTITY;
		createInfo.components.g = VK_COMPONENT_SWIZZLE_IDENTITY;
		createInfo.components.b = VK_COMPONENT_SWIZZLE_IDENTITY;
		createInfo.components.a = VK_COMPONENT_SWIZZLE_IDENTITY;
		createInfo.subresourceRange.aspectMask = VK_IMAGE_ASPECT_COLOR_BIT;
		createInfo.subresourceRange.baseMipLevel = 0;
		createInfo.subresourceRange.levelCount = 1;
		createInfo.subresourceRange.baseArrayLayer = 0;
		createInfo.subresourceRange.layerCount = 1;

		VK_CHECK(vkCreateImageView(device.get(), &createInfo, nullptr, &imageViews_[i]));
	}
}

File: csharp/Renderer/Swapchain.cs
using Silk.NET.Vulkan;
using Silk.NET.Windowing;

namespace IdleL.Rendering;

// Owns the swapchain, its images + views, the depth image, and the chosen format/extent.
// Rebuilds itself on resize via Recreate(). Uses (not owns) a GpuDevice.
unsafe class Swapchain : IDisposable
{
    readonly GpuDevice device;
    readonly Surface surface;
    readonly IWindow window;
    readonly Vk vk;

    public SwapchainKHR Handle;
    public Image[] Images = Array.Empty<Image>();
    public ImageView[] ImageViews = Array.Empty<ImageView>();
    public Extent2D Extent;
    public Format ImageFormat;
    public Format DepthFormat = Format.D32Sfloat;
    // one depth target per swapchain image so frames in flight don't share (and race) a single depth buffer
    public Image[] DepthImages = Array.Empty<Image>();
    public ImageView[] DepthViews = Array.Empty<ImageView>();
    Allocation[] depthMemories = Array.Empty<Allocation>();

    public SampleCountFlags Samples = SampleCountFlags.Count1Bit;
    public Image[] ColorImages = Array.Empty<Image>();        // multisampled render targets (per image), only when Samples != 1
    public ImageView[] ColorViews = Array.Empty<ImageView>();
    Allocation[] colorMemories = Array.Empty<Allocation>();

    public uint ImageCount => (uint)Images.Length;

    public Swapchain(GpuDevice device, Surface surface, IWindow window, SampleCountFlags samples)
    {
        this.device = device;
        this.surface = surface;
        this.window = window;
        Samples = samples;
        vk = device.Vk;
        CreateSwapchain();
        CreateImageViews();
        CreateColorResources();
        CreateDepthResources();
    }

    static SurfaceFormatKHR ChooseFormat(SurfaceFormatKHR[] formats)
    {
        foreach (var f in formats)
            if (f.Format == Format.B8G8R8A8Srgb && f.ColorSpace == ColorSpaceKHR.SpaceSrgbNonlinearKhr)
                return f;
        return formats[0];
    }

    static PresentModeKHR ChoosePresentMode(PresentModeKHR[] modes)
    {
        foreach (var m in modes) if (m == PresentModeKHR.MailboxKhr) return m;
        return PresentModeKHR.FifoKhr;
    }

    Extent2D ChooseExtent(SurfaceCapabilitiesKHR caps)
    {
        if (caps.CurrentExtent.Width != uint.MaxValue) return caps.CurrentExtent;
        var fb = window.FramebufferSize;
        return new Extent2D
        {
            Width = Math.Clamp((uint)fb.X, caps.MinImageExtent.Width, caps.MaxImageExtent.Width),
            Height = Math.Clamp((uint)fb.Y, caps.MinImageExtent.Height, caps.MaxImageExtent.Height)
        };
    }

    void CreateSwapchain()
    {
        var support = device.QuerySwapchainSupport(device.Physical);
        var format = ChooseFormat(support.Formats);
        var present = ChoosePresentMode(support.PresentModes);
        var extent = ChooseExtent(support.Capabilities);

        uint imageCount = support.Capabilities.MinImageCount + 1;
        if (support.Capabilities.MaxImageCount > 0 && imageCount > support.Capabilities.MaxImageCount)
            imageCount = support.Capabilities.MaxImageCount;

        var info = new SwapchainCreateInfoKHR
        {
            SType = StructureType.SwapchainCreateInfoKhr,
            Surface = surface.Handle,
            MinImageCount = imageCount,
            ImageFormat = format.Format,
            ImageColorSpace = format.ColorSpace,
            ImageExtent = extent,
            ImageArrayLayers = 1,
            ImageUsage = ImageUsageFlags.ColorAttachmentBit,
            PreTransform = support.Capabilities.CurrentTransform,
            CompositeAlpha = CompositeAlphaFlagsKHR.OpaqueBitKhr,
            PresentMode = present,
            Clipped = true,
            OldSwapchain = default
        };

        uint* families = stackalloc uint[] { device.GraphicsFamily, device.PresentFamily };
        if (device.GraphicsFamily != device.PresentFamily)
        {
            info.ImageSharingMode = SharingMode.Concurrent;
            info.QueueFamilyIndexCount = 2;
            info.PQueueFamilyIndices = families;
        }
        else info.ImageSharingMode = SharingMode.Exclusive;

        VkCheck.Check(device.KhrSwapchain.CreateSwapchain(device.Device, in info, null, out Handle), "vkCreateSwapchain");

        uint count = 0;
        device.KhrSwapchain.GetSwapchainImages(device.Device, Handle, ref count, null);
        Images = new Image[count];
        fixed (Image* p = Images) device.KhrSwapchain.GetSwapchainImages(device.Device, Handle, ref count, p);

        ImageFormat = format.Format;
        Extent = extent;
    }

    void CreateImageViews()
    {
        ImageViews = new ImageView[Images.Length];
        for (int i = 0; i < Images.Length; i++)
        {
            var info = new ImageViewCreateInfo
            {
                SType = StructureType.ImageViewCreateInfo,
                Image = Images[i],
                ViewType = ImageViewType.Type2D,
                Format = ImageFormat,
                Components = new ComponentMapping(ComponentSwizzle.Identity, ComponentSwizzle.Identity,
                                                  ComponentSwizzle.Identity, ComponentSwizzle.Identity),
                SubresourceRange = new ImageSubresourceRange(ImageAspectFlags.ColorBit, 0, 1, 0, 1)
            };
            VkCheck.Check(vk.CreateImageView(device.Device, in info, null, out ImageViews[i]), "swapchain view");
        }
    }

    // Multisampled color render target (resolved to the swapchain image at end-of-pass). Skipped
    // entirely when MSAA is off, then we render straight into the swapchain image.
    void CreateColorResources()
    {
        if (Samples == SampleCountFlags.Count1Bit) return;
        int n = Images.Length;
        ColorImages = new Image[n];
        ColorViews = new ImageView[n];
        colorMemories = new Allocation[n];
        for (int i = 0; i < n; i++)
        {
            device.CreateImage(Extent.Width, Extent.Height, 1, ImageFormat,
                ImageUsageFlags.ColorAttachmentBit, out ColorImages[i], out colorMemories[i], Samples);
            var view = new ImageViewCreateInfo
            {
                SType = StructureType.ImageViewCreateInfo,
                Image = ColorImages[i],
                ViewType = ImageViewType.Type2D,
                Format = ImageFormat,
                SubresourceRange = new ImageSubresourceRange(ImageAspectFlags.ColorBit, 0, 1, 0, 1)
            };
            VkCheck.Check(vk.CreateImageView(device.Device, in view, null, out ColorViews[i]), "msaa color view");
        }
    }

    void CreateDepthResources()
    {
        int n = Images.Length;
        DepthImages = new Image[n];
        DepthViews = new ImageView[n];
        depthMemories = new Allocation[n];
        for (int i = 0; i < n; i++)
        {
            device.CreateImage(Extent.Width, Extent.Height, 1, DepthFormat,
                ImageUsageFlags.DepthStencilAttachmentBit | ImageUsageFlags.SampledBit, out DepthImages[i], out depthMemories[i], Samples);
            var view = new ImageViewCreateInfo
            {
                SType = StructureType.ImageViewCreateInfo,
                Image = DepthImages[i],
                ViewType = ImageViewType.Type2D,
                Format = DepthFormat,
                SubresourceRange = new ImageSubresourceRange(ImageAspectFlags.DepthBit, 0, 1, 0, 1)
            };
            VkCheck.Check(vk.CreateImageView(device.Device, in view, null, out DepthViews[i]), "depth view");
        }
    }

    public void Recreate()
    {
        var fb = window.FramebufferSize;
        while (fb.X == 0 || fb.Y == 0) { fb = window.FramebufferSize; window.DoEvents(); }

        vk.DeviceWaitIdle(device.Device);
        Cleanup();
        CreateSwapchain();
        CreateImageViews();
        CreateColorResources();
        CreateDepthResources();
    }

    void Cleanup()
    {
        for (int i = 0; i < DepthImages.Length; i++)
        {
            vk.DestroyImageView(device.Device, DepthViews[i], null);
            vk.DestroyImage(device.Device, DepthImages[i], null);
            device.Allocator.Free(depthMemories[i]);
        }
        if (Samples != SampleCountFlags.Count1Bit)
            for (int i = 0; i < ColorImages.Length; i++)
            {
                vk.DestroyImageView(device.Device, ColorViews[i], null);
                vk.DestroyImage(device.Device, ColorImages[i], null);
                device.Allocator.Free(colorMemories[i]);
            }
        foreach (var v in ImageViews) vk.DestroyImageView(device.Device, v, null);
        device.KhrSwapchain.DestroySwapchain(device.Device, Handle, null);
    }

    public void Dispose() => Cleanup();
}
